File: creational-design-pattern/abstract-pattern/abstract_factory.go
// Abstract Factory Pattern
// ========================
// Category: Creational Design Pattern
//
// Intent:
//   Provide an interface for creating FAMILIES of related objects without
//   specifying their concrete types. Swapping the factory swaps the entire
//   product family.
//
// When to use:
//   - When your system needs families of related objects (e.g., themed UI)
//   - When you want to guarantee products from one family aren't mixed with another
//   - When you want to swap entire product families at runtime
//
// Key Participants:
//   - AbstractFactory (interface): Declares creation methods for each product
//   - ConcreteFactory: Implements creation for a specific family
//   - AbstractProduct (interface): Contract for each kind of product
//   - ConcreteProduct: Family-specific implementations

package main

import "fmt"

// ---------------------------------------------------------------------------
// Abstract Products (interfaces)
// ---------------------------------------------------------------------------

type Button interface {
	Render() string
	OnClick(handler string) string
}

type Checkbox interface {
	Render() string
	Toggle() string
}

type TextField interface {
	Render() string
	SetValue(value string) string
}

// ---------------------------------------------------------------------------
// Abstract Factory (interface)
// ---------------------------------------------------------------------------

type UIFactory interface {
	CreateButton() Button
	CreateCheckbox() Checkbox
	CreateTextField() TextField
}

// ---------------------------------------------------------------------------
// Dark Theme Family
// ---------------------------------------------------------------------------

type DarkButton struct{}

func (d *DarkButton) Render() string {
	return "<button class='bg-gray-800 text-white'>Click me</button>"
}
func (d *DarkButton) OnClick(handler string) string {
	return fmt.Sprintf("Dark button triggers: %s", handler)
}

type DarkCheckbox struct{}

func (d *DarkCheckbox) Render() string {
	return "<input type='checkbox' class='border-gray-600 bg-gray-700' />"
}
func (d *DarkCheckbox) Toggle() string {
	return "Dark checkbox toggled with smooth transition"
}

type DarkTextField struct{}

func (d *DarkTextField) Render() string {
	return "<input type='text' class='bg-gray-800 text-white' />"
}
func (d *DarkTextField) SetValue(value string) string {
	return fmt.Sprintf("Dark text field set to: %s", value)
}

type DarkThemeFactory struct{}

func (f *DarkThemeFactory) CreateButton() Button     { return &DarkButton{} }
func (f *DarkThemeFactory) CreateCheckbox() Checkbox  { return &DarkCheckbox{} }
func (f *DarkThemeFactory) CreateTextField() TextField { return &DarkTextField{} }

// ---------------------------------------------------------------------------
// Light Theme Family
// ---------------------------------------------------------------------------

type LightButton struct{}

func (l *LightButton) Render() string {
	return "<button class='bg-white text-black'>Click me</button>"
}
func (l *LightButton) OnClick(handler string) string {
	return fmt.Sprintf("Light button triggers: %s", handler)
}

type LightCheckbox struct{}

func (l *LightCheckbox) Render() string {
	return "<input type='checkbox' class='border-gray-300 bg-white' />"
}
func (l *LightCheckbox) Toggle() string {
	return "Light checkbox toggled with crisp animation"
}

type LightTextField struct{}

func (l *LightTextField) Render() string {
	return "<input type='text' class='bg-white text-black' />"
}
func (l *LightTextField) SetValue(value string) string {
	return fmt.Sprintf("Light text field set to: %s", value)
}

type LightThemeFactory struct{}

func (f *LightThemeFactory) CreateButton() Button     { return &LightButton{} }
func (f *LightThemeFactory) CreateCheckbox() Checkbox  { return &LightCheckbox{} }
func (f *LightThemeFactory) CreateTextField() TextField { return &LightTextField{} }

// ---------------------------------------------------------------------------
// Client Code — works ONLY through interfaces
// ---------------------------------------------------------------------------

func buildLoginForm(factory UIFactory) {
	username := factory.CreateTextField()
	password := factory.CreateTextField()
	remember := factory.CreateCheckbox()
	submit := factory.CreateButton()

	fmt.Println("  Username:", username.Render())
	fmt.Println("  Password:", password.Render())
	fmt.Println("  Remember:", remember.Render())
	fmt.Println("  Submit:  ", submit.Render())
	fmt.Println("  Action:  ", submit.OnClick("submit_login()"))
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	preference := "dark" // could come from config

	// Factory selection is the ONLY place concrete types appear
	var factory UIFactory
	if preference == "dark" {
		factory = &DarkThemeFactory{}
	} else {
		factory = &LightThemeFactory{}
	}

	fmt.Printf("Building login form with %s theme:\n", preference)
	buildLoginForm(factory)

	fmt.Println("\nBuilding login form with light theme:")
	buildLoginForm(&LightThemeFactory{})
}

File: creational-design-pattern/abstract-pattern/abstract_factory.rs
//! Abstract Factory Pattern
//! ========================
//! Category: Creational Design Pattern
//!
//! Intent:
//!   Provide an interface (trait) for creating FAMILIES of related objects
//!   without specifying their concrete types. Swapping the factory swaps
//!   the entire product family.
//!
//! When to use:
//!   - When your system needs families of related objects (e.g., themed UI)
//!   - When you want to guarantee that products from one family aren't mixed
//!     with products from another
//!
//! Key Participants:
//!   - AbstractFactory (trait): Declares creation methods for each product
//!   - ConcreteFactory: Implements creation for a specific family
//!   - AbstractProduct (trait): Interface for each kind of product
//!   - ConcreteProduct: Family-specific implementations

// ---------------------------------------------------------------------------
// Abstract Products (traits)
// ---------------------------------------------------------------------------
trait Button {
    fn render(&self) -> String;
    fn on_click(&self, handler: &str) -> String;
}

trait Checkbox {
    fn render(&self) -> String;
    fn toggle(&self) -> String;
}

trait TextField {
    fn render(&self) -> String;
    fn set_value(&self, value: &str) -> String;
}

// ---------------------------------------------------------------------------
// Abstract Factory (trait)
// ---------------------------------------------------------------------------
trait UIFactory {
    fn create_button(&self) -> Box<dyn Button>;
    fn create_checkbox(&self) -> Box<dyn Checkbox>;
    fn create_text_field(&self) -> Box<dyn TextField>;
}

// ---------------------------------------------------------------------------
// Dark Theme Family
// ---------------------------------------------------------------------------
struct DarkButton;
impl Button for DarkButton {
    fn render(&self) -> String {
        "<button class='bg-gray-800 text-white'>Click me</button>".to_string()
    }
    fn on_click(&self, handler: &str) -> String {
        format!("Dark button triggers: {}", handler)
    }
}

struct DarkCheckbox;
impl Checkbox for DarkCheckbox {
    fn render(&self) -> String {
        "<input type='checkbox' class='border-gray-600 bg-gray-700' />".to_string()
    }
    fn toggle(&self) -> String {
        "Dark checkbox toggled with smooth transition".to_string()
    }
}

struct DarkTextField;
impl TextField for DarkTextField {
    fn render(&self) -> String {
        "<input type='text' class='bg-gray-800 text-white' />".to_string()
    }
    fn set_value(&self, value: &str) -> String {
        format!("Dark text field set to: {}", value)
    }
}

struct DarkThemeFactory;
impl UIFactory for DarkThemeFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(DarkButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(DarkCheckbox)
    }
    fn create_text_field(&self) -> Box<dyn TextField> {
        Box::new(DarkTextField)
    }
}

// ---------------------------------------------------------------------------
// Light Theme Family
// ---------------------------------------------------------------------------
struct LightButton;
impl Button for LightButton {
    fn render(&self) -> String {
        "<button class='bg-white text-black'>Click me</button>".to_string()
    }
    fn on_click(&self, handler: &str) -> String {
        format!("Light button triggers: {}", handler)
    }
}

struct LightCheckbox;
impl Checkbox for LightCheckbox {
    fn render(&self) -> String {
        "<input type='checkbox' class='border-gray-300 bg-white' />".to_string()
    }
    fn toggle(&self) -> String {
        "Light checkbox toggled with crisp animation".to_string()
    }
}

struct LightTextField;
impl TextField for LightTextField {
    fn render(&self) -> String {
        "<input type='text' class='bg-white text-black' />".to_string()
    }
    fn set_value(&self, value: &str) -> String {
        format!("Light text field set to: {}", value)
    }
}

struct LightThemeFactory;
impl UIFactory for LightThemeFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(LightButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(LightCheckbox)
    }
    fn create_text_field(&self) -> Box<dyn TextField> {
        Box::new(LightTextField)
    }
}

// ---------------------------------------------------------------------------
// Client Code — works ONLY through trait objects
// ---------------------------------------------------------------------------
fn build_login_form(factory: &dyn UIFactory) {
    let username = factory.create_text_field();
    let password = factory.create_text_field();
    let remember = factory.create_checkbox();
    let submit = factory.create_button();

    println!("  Username: {}", username.render());
    println!("  Password: {}", password.render());
    println!("  Remember: {}", remember.render());
    println!("  Submit:   {}", submit.render());
    println!("  Action:   {}", submit.on_click("submit_login()"));
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    let preference = "dark"; // could come from config

    // The factory selection is the ONLY place concrete types appear
    let factory: Box<dyn UIFactory> = match preference {
        "dark" => Box::new(DarkThemeFactory),
        _ => Box::new(LightThemeFactory),
    };

    println!("Building login form with {} theme:", preference);
    build_login_form(&*factory);

    println!("\nBuilding login form with light theme:");
    build_login_form(&LightThemeFactory);
}
